tests/test_vercel_ai.py

```diff
@@ -1964,3 +1973,586 @@ async def test_adapter_load_messages():
+                role='assistant',
+                parts=[TextUIPart(text='First second', state='done')],
+            )
+        ]
+    )
+
+
+async def test_adapter_dump_messages_text_with_interruption():
+    """Test text concatenation with interruption."""
+    messages = [
+        ModelResponse(
+            parts=[
+                TextPart(content='Before tool'),
+                BuiltinToolCallPart(
+                    tool_name='test',
+                    args={},
+                    tool_call_id='t1',
+                    provider_name='test',
+                ),
+                BuiltinToolReturnPart(
+                    tool_name='test',
+                    content='result',
+                    tool_call_id='t1',
+                    provider_name='test',
+                ),
+                TextPart(content='After tool'),
+            ]
+        ),
+    ]
+
+    id_gen = predictable_id_generator()
+    ui_messages = VercelAIAdapter.dump_messages(messages, _id_generator=id_gen)
+
+    assert ui_messages == snapshot(
+        [
+            UIMessage(
+                id='test-id-1',
+                role='assistant',
+                parts=[
+                    TextUIPart(text='Before tool', state='done'),
+                    ToolOutputAvailablePart(
+                        type='tool-test',
+                        tool_call_id='pyd_ai_builtin|test|t1',
+                        input='{}',
+                        output='result',
+                        state='output-available',
+                        provider_executed=True,
+                        call_provider_metadata={'pydantic_ai': {'provider_name': 'test'}},
+                    ),
+                    TextUIPart(text='\n\nAfter tool', state='done'),
+                ],
+            )
+        ]
+    )
+
+
+async def test_adapter_dump_load_roundtrip():
+    """Test that dump_messages and load_messages are approximately inverse operations."""
+    original_messages = [
+        ModelRequest(
+            parts=[
+                SystemPromptPart(content='System message'),
+                UserPromptPart(content='User message'),
+            ]
+        ),
+        ModelResponse(
+            parts=[
+                TextPart(content='Response text'),
+                ToolCallPart(tool_name='tool1', args={'key': 'value'}, tool_call_id='tc1'),
+            ]
+        ),
+        ModelRequest(parts=[ToolReturnPart(tool_name='tool1', content='tool result', tool_call_id='tc1')]),
+        ModelResponse(
+            parts=[
+                TextPart(content='Final response'),
+            ]
+        ),
+    ]
+
+    id_gen = predictable_id_generator()
+    ui_messages = VercelAIAdapter.dump_messages(original_messages, _id_generator=id_gen)
+
+    # Load back to Pydantic AI format
+    reloaded_messages = VercelAIAdapter.load_messages(ui_messages)
+
+    # Can't use `assert reloaded_messages == original_messages` because the timestamps will be different
```

Comment:
What if we use a helper function that walks the list and resets the timestamps? `assert reloaded_messages == original_messages` would be much less brittle than having to manually review these 2 snapshots for equality